client/messaging/common-message-layout.tsx

```diff
@@ -34,55 +36,84 @@ const Text = styled.span`
   overflow: hidden;
 `
 
-export function ParsedText({ text }: { text: string }) {
-  const parsedText = useMemo(() => {
-    const matches = matchLinks(text)
+const UserMention = styled.span`
+  background-color: ${rgba(blue400, 0.48)};
+
+  &:hover {
+    background-color: ${rgba(blue400, 0.87)};
+  }
+`
+
+function* getAllMatches(text: string) {
+  yield* matchMentionsMarkup(text)
+  yield* matchLinks(text)
+}
+
+export const TextMessage = React.memo<{
+  userId: SbUserId
+  selfUserId: SbUserId
+  time: number
+  text: string
+}>(props => {
+  const { userId, selfUserId, time, text } = props
+  const [parsedText, isHighlighted] = useMemo(() => {
+    const matches = getAllMatches(text)
+    const sortedMatches = Array.from(matches).sort((a, b) => a.index - b.index)
     const elements = []
     let lastIndex = 0
+    let isHighlighted = false
 
-    for (const match of matches) {
+    for (const match of sortedMatches) {
```

Comment:
This code doesn't seem to handle the case that `match.index < lastIndex`, which seems necessary. Not sure if we don't have any tests that trigger this (it seems like you *tried* to have that, but maybe the mention pattern is restrictive enough that it didn't actually trigger like you thought). In any case, if this occurs, the match should be skipped entirely (as it overlaps with a previous match).

Thinking more about the mention pattern, I can't think of a way to test that currently, but this logic should exist so we don't get surprised in the future when we add other patterns.